### PR TITLE
Add ability for artifact install to create directories

### PR DESCRIPTION
**Pull request by sbaxter-izo**
This PR adds the ability to create directories (and nested directories) to the `mender-artifact install` command, mimicking the `install` utility syntax:
`mender-artifact install -d <artifact>:/dirs/to/create`

Changelog: Title

Signed-off-by: Sam Baxter <sbaxter@izotope.com>

### CODE REVIEW

File: cli/mender-artifact/debugfs.go [hunk 1]
```diff
@@ -73,6 +73,22 @@ func debugfsCopyFile(file, image string) (ret string, err error) {
...
 }
 
+func debugfsMakeDir(imageFile, image string) (err error) {
+	// Remove the `/` suffix if present, as debugfs does not play nice with it.
+	imageFile = strings.TrimRight(imageFile, "/")
+	// Recursively create parent directories if they do not exist
+	dir, _ := filepath.Split(imageFile)
+	_, err = os.Stat(dir)
+	if dir != "" && os.IsNotExist(err) {
+		debugfsMakeDir(dir, image)
```
Comment (kacf) [inline cli/mender-artifact/debugfs.go]: I don't quite follow this code. Isn't `imageFile` the path inside of the image? So why is `dir` being tested using `os.Stat` on the host filesystem?

File: cli/mender-artifact/main.go [hunk 2]
```diff
@@ -395,8 +395,8 @@ func getCliContext() *cli.App {
 
 	install := cli.Command{
 		Name:        "install",
-		Usage:       "install -m <permissions> <hostfile> [artifact|sdimg|uefiimg]:<filepath>",
-		Description: "Installs a file from the host filesystem to the artifact or sdimg.",
+		Usage:       "install -m <permissions> <hostfile> [artifact|sdimg|uefiimg]:<filepath> or install -d [artifact|sdimg|uefiimg]:<directory>",
+		Description: "Installs a file or directory from the host filesystem to the artifact or sdimg.",
```
Comment (kacf) [inline cli/mender-artifact/main.go]: Suggest rephrasing to `Installs a directory, or a file from the host filesystem, to the artifact or sdimg`.

File: cli/mender-artifact/main.go [hunk 3]
```diff
@@ -405,6 +405,10 @@ func getCliContext() *cli.App {
 			Name:  "mode, m",
 			Usage: "Set the permission bits in the file",
 		},
+		cli.BoolFlag{
+			Name:  "directory, d",
+			Usage: "Install a directory into an artifact",
```
Comment (kacf) [inline cli/mender-artifact/main.go]: I would rephrase as `Create a directory inside an artifact`.

File: cli/mender-artifact/debugfs.go (hunk 1 above)
Comment (sbaxter-izo) [reply cli/mender-artifact/debugfs.go]: Good catch! The UNIX `install -d` silently does nothing if the directory already exists, so I'll mirror that behavior.

File: cli/mender-artifact/main.go (hunk 2 above)
Comment (sbaxter-izo) [reply cli/mender-artifact/main.go]: Fixed the phrasing

File: cli/mender-artifact/main.go (hunk 3 above)
Comment (sbaxter-izo) [reply cli/mender-artifact/main.go]: Fixed the phrasing.

File: cli/mender-artifact/debugfs.go (hunk 1 above)
Comment (sbaxter-izo) [reply cli/mender-artifact/debugfs.go]: I removed this `os.Stat` logic and replaced it with a `debugfs stat` check to simply return if the directory already exists.

File: cli/mender-artifact/copy.go [hunk 4]
```diff
@@ -161,11 +161,24 @@ func Install(c *cli.Context) (err error) {
...
-		if c.Int("mode") == 0 {
+		if c.Int("mode") == 0 && !directory {
 			return cli.NewExitError("File permissions needs to be set, if you are simply copying, the cp command should fit your needs", 1)
 		}
 		perm = os.FileMode(c.Int("mode"))
+		if directory {
+			vdir, err := virtualImage.OpenDir(comp, privateKey, c.Args().First())
+			defer wclose(vdir)
+			if err != nil {
+				return cli.NewExitError(fmt.Sprintf("%v", err), 1)
```
[the remark this answers is not in the document]
Comment (sbaxter-izo) [reply cli/mender-artifact/copy.go]: Removed `fmt.Sprintf`

**Comment by sbaxter-izo** [comment]
Thanks for the review!

> Do I understand correctly that it works more or less like the Unix `install -d` tool? This means that the directory does not necessarily have to exist on the build host filesystem, correct?

Yep, that is the intention and I've resoled the requested changes. Ready for another look.

File: cli/mender-artifact/debugfs.go [hunk 5]
```diff
@@ -74,12 +74,16 @@ func debugfsCopyFile(file, image string) (ret string, err error) {
 }
 
 func debugfsMakeDir(imageFile, image string) (err error) {
+	_, err = debugfsExecuteCommand(fmt.Sprintf("stat %s", imageFile), image)
```
Comment (kacf) [inline cli/mender-artifact/debugfs.go:77]: Should this also go after the `TrimRight`?

Comment (sbaxter-izo) [reply cli/mender-artifact/debugfs.go:77]: `debugfs stat` doesn't mind the trailing `/`, but `mkdir` does. I've clarified the comment to say it is needed for the `mkdir` command

**Comment by sbaxter-izo** [comment]
Ah, darn, I forgot that. Fixed now.

Thanks for the review! Every time I've interacted with the mender community I've received great support, happy to contribute.

**Comment by sbaxter-izo** [comment]
@kacf Is there anything with the failing checks that I need to resolve so that this can get merged?

I'm also curious how/when this upstream work gets integrated into the `meta-mender` upstream. We're upgrading from an older poky to latest zeus. I've noticed there doesn't seem to be zeus support in `meta-mender` upstream yet either, but I'm wondering if this change will make it in there once zeus support is complete.

Thanks again!

**Comment by kacf** [comment]
It appears that the current failure is due to some unrelated network timeout, so I'm rerunning that one.

This will end up both in zeus and in older branches, don't worry about that! zeus doesn't have its own branch yet because we've had a lot of integration problems with it, but it is being worked on.

**Comment by sbaxter-izo** [comment]
Great to hear!

**Comment by kacf** [comment]
Thanks @sbaxter-izo!
